Edits from final review (#37)
OpenLiberty/guide-microprofile-config-profile: README.adoc

:page-layout: guide-multipane
:page-duration: 20 minutes
:page-releasedate: 2024-02-29
:page-description: Learn how to use MicroProfile Config's configuration profiles to externalize configurations for different phases of the CI/CD lifecycle.
:page-tags: ['microprofile']
:page-permalink: /guides/{projectid}
:imagesdir: /img/guide/{projectid}
:page-related-guides: ['microprofile-config-intro', 'microprofile-config']
:common-includes: https://raw.githubusercontent.com/OpenLiberty/guides-common/prod
:source-highlighter: prettify
:page-seo-title: Configuring Java microservices for multiple environments using MicroProfile Config

> :page-seo-description: A tutorial and example on how to externalize configuration of Java microservices and applications for different deployment environments in the CI/CD lifecycle using MicroProfile Config configuration profiles.

Reviewer: ```suggestion
:page-seo-description: A tutorial on how to use MicroProfile Config configuration profiles to externalize configuration of Java microservices and applications for different deployment environments in the CI/CD lifecycle. 
```

:guide-author: Open Liberty
= Externalizing environment-specific microservice configuration for CI/CD

[.hidden]
NOTE: This repository contains the guide documentation source. To view the guide in published form, view it on the https://openliberty.io/guides/{projectid}.html[Open Liberty website].

> Learn how to create environment-specific configurations for microservices by using MicroProfile Config's configuration profile for easy management and portable deployments throughout the CI/CD lifecycle.

Reviewer: ```suggestion
Learn how to create environment-specific configurations for microservices by using MicroProfile Config configuration profiles for easy management and portable deployments throughout the CI/CD lifecycle.
```

// =================================================================================================
== What you'll learn

> In the domain of software delivery, managing configurations for microservices can become complex and challenging, especially when configurations require adjustments across various stages of the software development and delivery lifecycle. MicroProfile Config's configuration profile feature, also known as the https://download.eclipse.org/microprofile/microprofile-config-3.0/microprofile-config-spec-3.0.html#configprofile[Config Profile^], is a direct solution to this challenge. It simplifies the management of microservice configurations across diverse environments - from development to production and throughout the CI/CD (Continuous Integration/Continuous Delivery) pipeline. By externalizing and tailoring configuration properties to each environment, the CI/CD process becomes more seamless, allowing developers to concentrate on perfecting their application's code and capabilities.

Reviewer: ```suggestion
Managing configurations for microservices can be challenging, especially when configurations require adjustments across various stages of the software development and delivery lifecycle. The MicroProfile Config configuration profile feature, also known as the https://download.eclipse.org/microprofile/microprofile-config-3.0/microprofile-config-spec-3.0.html#configprofile[Config Profile^], is a direct solution to this challenge. It simplifies the management of microservice configurations across diverse environments - from development to production and throughout the  continuous integration/continuous delivery (CI/CD) pipeline. By externalizing and tailoring configuration properties to each environment, the CI/CD process becomes more seamless, so you can concentrate on perfecting your application code and capabilities.
```

> You'll learn how to provide environment-specific configurations by using MicroProfile Config's configuration profile feature. You'll create configuration profiles using individual profile-specific configuration properties and profile-specific default `microprofile-config.properties` configuration sources, working with the MicroProfile Config API.

Reviewer: ```suggestion
You'll learn how to provide environment-specific configurations by using the MicroProfile Config configuration profile feature. You'll work with the MicroProfile Config API to create configuration profiles that use profile-specific configuration properties and configuration sources.
```

> This guide builds on the basics of the https://openliberty.io/guides/microprofile-config-intro.html[Separating configuration from code in microservices^] guide and the https://openliberty.io/guides/microprofile-config.html[Configuring microservices^] guide. If you are not familiar with externalizing the configuration of microservices, it will be helpful to read the https://openliberty.io/docs/latest/external-configuration.html[External configuration of microservices^] document and complete the aforementioned guides before proceeding.

Reviewer: ```suggestion
This guide builds on the https://openliberty.io/guides/microprofile-config-intro.html[Separating configuration from code in microservices^] guide and the https://openliberty.io/guides/microprofile-config.html[Configuring microservices^] guide. If you are not familiar with externalizing the configuration of microservices, it will be helpful to read the https://openliberty.io/docs/latest/external-configuration.html[External configuration of microservices^] document and complete the aforementioned guides before you proceed.
```

> The application that you will work with is a `query` service, which fetches information about the running JVM from a `system` microservice. Given the differences in setup between development, testing, and production environments, you'll use MicroProfile Config's configuration profiles to externalize and manage the configurations across the different environments.

Reviewer: ```suggestion
The application that you will work with is a `query` service, which fetches information about the running JVM from a `system` microservice. You'll use configuration profiles to externalize and manage the configurations across the development, testing, and production environments.
```

// =================================================================================================
== Creating a configuration profile for the dev environment

> The dev environment is used in the development stage by developers to test, experiment, debug, and refine their code, ensuring an application's functional readiness before progressing to subsequent stages in a software development and delivery lifecycle.

Reviewer: ```suggestion
The dev environment is used to test, experiment, debug, and refine your code, ensuring an application's functional readiness before progressing to subsequent stages in a software development and delivery lifecycle.
```

include::finish/query/pom.xml[]
----

> The `system` microservice contains the three Maven build profiles: [hotspot=development file=0]`dev`, [hotspot=testing file=0]`test`, and [hotspot=prod file=0]`prod`, in which the [hotspot=defaultProfile file=0]`dev` profile is set as the default. Each build profile defines properties for a particular deployment configuration that the microservice should use.

Reviewer: ```suggestion
The `system` microservice contains the three Maven build profiles: [hotspot=development file=0]`dev`, [hotspot=testing file=0]`test`, and [hotspot=prod file=0]`prod`, in which the [hotspot=defaultProfile file=0]`dev` profile is set as the default. Each build profile defines properties for a particular deployment configuration that the microservice uses.
```

> MicroProfile Config's configuration profile feature allows for the supply of configurations for different environments while only a single profile is active. The active profile is set using the `mp.config.profile` property. It can be set in any of the https://openliberty.io/docs/latest/external-configuration.html#default[configuration sources^] and is read once during application startup. When a profile is active, its associated configuration properties are used. For the `query` service, the `mp.config.profile` property is set to [hotspot=mp.config.profile file=1]`dev` in its Maven `pom.xml` as a Liberty configuration variable indicating to the runtime that `dev` is the active configuration profile.

Reviewer: ```suggestion
The MicroProfile Config configuration profile feature supplies configurations for different environments while only a single profile is active. The active profile is set using the `mp.config.profile` property. You can set it in any of the https://openliberty.io/docs/latest/external-configuration.html#default[configuration sources^] and it is read once during application startup. When a profile is active, its associated configuration properties are used. For the `query` service, the `mp.config.profile` property is set to [hotspot=mp.config.profile file=1]`dev` in its Maven `pom.xml`. This Liberty configuration variable indicates to the runtime that `dev` is the active configuration profile.
```

----

In the dev environment, the [hotspot=development file=0]`dev` configuration profile is set in the [hotspot file=0]`system/pom.xml` file as the configuration profile to use for running the `system` service. The `system` service runs on HTTP port [hotspot=httpport file=0]`9081` and HTTPS port [hotspot=httpsport file=0]`9444` using the context root [hotspot=context.root file=0]`system/dev`. It uses a basic user registry with username [hotspot=username file=0]`alice` and password [hotspot=password file=0]`alicepwd` for resource authorization. Note that the `basicRegistry` element is a simple registry configuration for learning purposes. For more information on user registries, see the https://openliberty.io/docs/latest/user-registries-application-security.html[User registries documentation^].

Point your browser to the http://localhost:9085/query/systems/localhost URL. The `query` service returns the message: `{"fail":"Failed to reach the client localhost."}`. This is because the current `query` service uses the default properties in the [hotspot file=2]`query/src/main/resources/META-INF/microprofile-config.properties` file to access the `system` service.

> For proper communication with the development `system` service, the `query` service should use the properties in the `dev` configuration profile.

Reviewer: ```suggestion
For proper communication with the development `system` service, the `query` service uses the properties in the `dev` configuration profile.
```

image::system-query-devops-development.png[System service running in development environment,align="center",width=85%,height=85%]

There are two ways to define configuration properties associated with your configuration profile. The first is as individual configuration properties associated with a configuration profile that can be specified in any kind of MicroProfile configuration source. The second is through default `microprofile-config.properties` configuration files embedded inside your application that can be associated with different configuration profiles. The former allows for flexibility in defining profile-specific configuration properties in the best configuration sources for your needs while the latter enables default profiles of configuration properties to be provided in your application.

// =================================================================================================
// Configuring properties at the property level
// =================================================================================================

> === Creating individual profile-specific configuration properties

Reviewer: ```suggestion
=== Creating profile-specific configuration properties
```

Configure the [hotspot=development file=0]`%dev.*` properties in the `microprofile-config.properties` file based on the values from the [hotspot=developmentProperties file=1]`dev` profile of the `system` service.

> Because the active profile is set to `dev`, each `%dev.*` property will override the value of the plain non-profile-specific property. For example, the [hotspot=dev.port file=0]`%dev.system.httpsPort` property will override the [hotspot=port file=0]`system.httpsPort` property and the value will be resolved to `9444` in this case.

Reviewer: ```suggestion
Because the active profile is set to `dev`, each `%dev.*` property overrides the value of the plain non-profile-specific property. For example, in this case the [hotspot=dev.port file=0]`%dev.system.httpsPort` property overrides the [hotspot=port file=0]`system.httpsPort` property and the value is resolved to `9444`.
```

Because you are running the `query` service in dev mode, the changes that you made are automatically picked up. Try out the application at the http://localhost:9085/query/systems/localhost URL. You can see the current OS and Java version in JSON format.

// =================================================================================================
// Configuring properties using higher-level configuration sources
// =================================================================================================
=== Creating profile-specific `microprofile-config.properties` configuration files

> Creating profile-specific offers `microprofile-config.properties` configuration files offers a structured way to provide and manage more extensive sets of default configurations. You can create a configuration file for each configuration profile using the naming convention `microprofile-config-<config_profile_id>` in the `META-INF` folder on the classpath of your application, where `<config_profile_id>` is the unique identifier for a configuration profile. Once you have created the file, you can add your configuration properties to it with the standard `<property_name>=<value>` syntax.

Reviewer: ```suggestion
Creating profile-specific `microprofile-config.properties` configuration files is a structured way to provide and manage more extensive sets of default configurations. You can create a configuration file for each configuration profile in the `META-INF` folder on the classpath of your application by using the `microprofile-config-<config_profile_id>` naming convention, where `<config_profile_id>` is the unique identifier for a configuration profile. Once you create the file, you can add your configuration properties to it with the standard `<property_name>=<value>` syntax.
```

Remove the `%dev.*` properties from the `microprofile-config.properties` file.

> Because the active profile is set to `dev`, the [hotspot file=0]`microprofile-config-dev.properties` file is loaded on top of the default [hotspot file=2]`microprofile-config.properties` file. Any [hotspot=system file=0]`system.*` properties specified in the `microprofile-config-dev.properties` file takes precedence over the [hotspot=system file=2]`system.*` property values in the `microprofile-config.properties` file.

Reviewer: ```suggestion
Because the active profile is set to `dev`, any [hotspot=system file=0]`system.*` properties specified in the `microprofile-config-dev.properties` file take precedence over the [hotspot=system file=2]`system.*` property values in the `microprofile-config.properties` file.
```

// =================================================================================================
== Creating a configuration profile for the test environment

> In CI/CD, the test environment is where integration tests come to life to ensure readiness and quality of an application. A good testing configuration not only ensures smooth operations but also aligns the environment closely with potential production settings.

Reviewer: ```suggestion
In CI/CD, the test environment is where integration tests ensure the readiness and quality of an application. A good testing configuration not only ensures smooth operations but also aligns the environment closely with potential production settings.
```

// =================================================================================================
== Next steps

Deploying the application to a Kubernetes environment using the Open Liberty Operator is an optional learning step in this guide.

To further explore deploying microservices using Kubernetes and the Open Liberty Operator, you can read the following guides:

* https://openliberty.io/guides/openliberty-operator-intro.html[Deploying a microservice to Kubernetes using Open Liberty Operator^]
* https://openliberty.io/guides/openliberty-operator-openshift.html[Deploying a microservice to OpenShift 4 using Open Liberty Operator^]

> A secure production environment is essential to application security. In the previous sections, you learned how to externalize credentials and other properties for accessing the `system` service by using MicroProfile Config API. This strategy makes the application more adaptable to different environments without the need to change code and rebuild your application. 

Reviewer: ```suggestion
A secure production environment is essential to application security. In the previous sections, you learned how to use the MicroProfile Config API to externalize credentials and other properties for accessing the `system` service. This strategy makes the application more adaptable to different environments without the need to change code and rebuild your application. 
```
